### PR TITLE
feat(caldav): linkify location in scheduling mails

### LINKED ISSUE #5389: Improve handling of links for online events
### Is your feature request related to a problem? Please describe.

As a user of online meetings I want a smooth experience for links I have to open to join a call.

### Describe the solution you'd like

1. Write the conference URL into the location field when creating Talk rooms for an event. https://github.com/nextcloud/calendar/pull/5395
2. Implement a clickable link for the location field. https://github.com/nextcloud/calendar/pull/5355
3. Render the location field as a clickable link in the appointment notification mail (if it contains a link). https://github.com/nextcloud/server/pull/39753

### Describe alternatives you've considered

_No response_

### Additional context

This is an alternative approach to https://github.com/nextcloud/calendar/issues/1758 while it's poorly supported by other clients.

### PR DESCRIPTION
**Pull request by st3iny**
* Resolves: https://github.com/nextcloud/calendar/issues/5389

### Invitation
grafik

### Event is updated
grafik

## TODO

- [ ] Probably tests

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Comment by st3iny** [comment]
I tried to inject evil code into the location field but it was sanitized properly. The general email html sanitization seems to take care of this.

File: apps/dav/lib/CalDAV/Schedule/IMipService.php [hunk 1]
```diff
@@ -105,6 +106,33 @@ private function generateDiffString(VEvent $vevent, VEvent $oldVEvent, string $p
...
+
+	/**
+	 * Convert a given url to a html link element or return null otherwise.
+	 */
+	private function linkify(string $url): ?string {
+		if (!str_starts_with($url, 'http://') && !str_starts_with($url, 'https://')) {
+			return null;
+		}
+
+		return sprintf('<a href="%1$s">%1$s</a>', $url);
```
[the remark this answers is not in the document]
Comment (st3iny) [reply apps/dav/lib/CalDAV/Schedule/IMipService.php:133]: Good point. I added a call to `htmlspecialchars`.

**Review by ChristophWurst — APPROVED** [review]
Looks good!

File: apps/dav/lib/CalDAV/Schedule/IMipService.php [hunk 2]
```diff
@@ -105,6 +106,33 @@ private function generateDiffString(VEvent $vevent, VEvent $oldVEvent, string $p
...
+			$oldstring = $oldVEvent->$property->getValue();
+			return sprintf($strikethrough, $this->linkify($oldstring) ?? $oldstring, $this->linkify($newstring) ?? $newstring);
+		}
+		return $this->linkify($newstring) ?? $newstring;
+	}
+
+	/**
+	 * Convert a given url to a html link element or return null otherwise.
+	 */
+	private function linkify(string $url): ?string {
```
[the remark this answers is not in the document]
Comment (ChristophWurst) [reply apps/dav/lib/CalDAV/Schedule/IMipService.php]: ``\Sabre\VObject\Property::getValue`` lies and says it only returns strings. I'll send a fix upstream.

[the remark this answers is not in the document]
Comment (ChristophWurst) [reply apps/dav/lib/CalDAV/Schedule/IMipService.php]: https://github.com/sabre-io/vobject/commit/87037b2ba489d6a5178e23475414c5f3280272da well of course you are a year ahead of us

File: apps/dav/lib/CalDAV/Schedule/IMipService.php [hunk 3]
```diff
@@ -105,6 +106,37 @@
...
+	private function generateLinkifiedDiffString(VEvent $vevent, VEvent $oldVEvent, string $property, string $default): ?string {
+		$strikethrough = "<span style='text-decoration: line-through'>%s</span><br />%s";
+		if (!isset($vevent->$property)) {
+			return $default;
+		}
+		/** @var string|null $newstring */
+		$newstring = $vevent->$property->getValue();
+		if (isset($oldVEvent->$property) && $oldVEvent->$property->getValue() !== $newstring) {
+			$oldstring = $oldVEvent->$property->getValue();
+			return sprintf($strikethrough, $this->linkify($oldstring) ?? $oldstring, $this->linkify($newstring) ?? $newstring);
```
[the remark this answers is not in the document]
Comment (tcitworld) [reply apps/dav/lib/CalDAV/Schedule/IMipService.php:121]: Maybe handle this as well?

**Review by tcitworld — APPROVED** [review]

**Pull request merged**